cirq/ops/common_channels.py

```diff
@@ -187,3 +189,528 @@ def depolarize(p: float) -> DepolarizingChannel:
         ValueError if p is not a valid probability.
     """
     return DepolarizingChannel(p)
+
+
+class GeneralizedAmplitudeDampingChannel(raw_types.Gate):
+    """ The generalized amplitude damping channel."""
+
+    def __init__(self, p, gamma) -> None:
+
+        """
+        This channel models the effect of dissipation to the environment
+        at possibly non zero tempurature. Where gamma is the damping rate
+        and p is the probability of
+
+        This channel evovles a density matrix via
+            \rho -> M_0 \rho M_0^\dagger + M_1 \rho M_1^\dagger
+                  + M_2 \rho M_2^\dagger + M_3 \rho M_3^\dagger
+
+        With
+            M_0 = \sqrt{p} \begin{bmatrix} 1 & 0  \\
+                            0 & \sqrt{1 - \gamma} \end{bmatrix}
+            M_1 = \sqrt{p} \begin{bmatrix} 0 & \sqrt{\gamma} \\
+                            0 & 0 \end{bmatrix}
+            M_2 = \sqrt{1-p} \begin{bmatrix} \sqrt{1-\gamma} & 0 \\
+                            0 & 1 \end{bmatrix}
+            M_3 = \sqrt{1-p} \begin{bmatrix} 0 & 0 \\
+                            \sqrt{gamma} & 0 \end{bmatrix}
+
+        Args:
+            gamma: The damping rate
+            p: the probability of
+
+        Raises:
+            ValueError is gamma or p is not a valid probability.
+        """
+
+        def validate_probability(p, p_str):
+            if p < 0:
+                raise ValueError('{} was less than 0.'.format(p_str))
+            elif p > 1:
+                raise ValueError('{} was greater than 1.'.format(p_str))
+            return p
+
+        self._gamma = validate_probability(gamma, 'gamma')
+        self._p = validate_probability(p, 'p')
+
+    def _channel_(self) -> Iterable[np.ndarray]:
+        return (
+            np.sqrt(self._p)
```

Comment:
Precompute the `np.sqrt`s to avoid redoing work, and to make the matrices easier to read. E.g. `p0 = np.sqrt(self._p)`, `p1 = np.sqrt(1 - self._p)`.

Reply:
fixed